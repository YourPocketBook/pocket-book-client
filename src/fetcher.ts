import isFuture from "date-fns/isFuture";
import jwtDecode from "jwt-decode";

class Fetcher {
  private realName: string = "";
  private nameChanged?: (name: string) => void;
  private token: string | null;
  private parsedToken:
    | { exp: number; given_name: string; role: string }
    | undefined;

  constructor() {
    this.token = localStorage.getItem("token");
    this.parseToken();
  }

  public setNameChangeHandler(func?: (name: string) => void) {
    this.nameChanged = func;
  }

  public authFetch(input: RequestInfo, init?: RequestInit): Promise<Response> {
    if (!this.token) {
      throw new Error("No token");
    }

    init = {
      ...init,
      headers: {
        ...(init ? init.headers : {}),
        Authorization: `Bearer ${this.token}`
      }
    };

    return fetch(input, init);
  }

  public saveToken(token: string) {
    localStorage.setItem("token", token);
    this.token = token;
    this.parseToken();
  }

  public checkToken(token: string): boolean {
    return this.token === token;
  }

  public clearToken() {
    localStorage.removeItem("token");
    this.token = null;
    this.parsedToken = undefined;
  }

  public getName() {
    return this.realName;
  }

  public updateName(name: string) {
    this.realName = name;
    if (this.nameChanged) {
      this.nameChanged(name);
    }
  }

  public isAdmin() {
    if (this.parsedToken) {
      return this.parsedToken.role === "Admin";
    }
    return false;
  }

  public hasToken(): boolean {
    return !!this.token;
  }

  public isInDate() {
    if (this.parsedToken) {
      return isFuture(this.parsedToken.exp * 1000);
    }

    return false;
  }

  private parseToken() {
    if (this.token) {
      try {
        this.parsedToken = jwtDecode(this.token);
      } catch {
        this.token = null;
        this.updateName("");
        localStorage.removeItem("token");
        this.parsedToken = undefined;
      }

      if (this.parsedToken) {
        this.updateName(this.parsedToken.given_name);
      }
    } else {
      this.parsedToken = undefined;
      this.updateName("");
    }
  }
}

const fetcher = new Fetcher();

export { fetcher, Fetcher };
