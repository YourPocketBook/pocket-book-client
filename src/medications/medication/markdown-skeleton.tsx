import React from "react";
import Skeleton from "react-loading-skeleton";
import ReactMarkdown from "react-markdown";

export function MarkdownSkeleton({ source }: { source?: string }) {
  return source ? <ReactMarkdown source={source} /> : <Skeleton count={3} />;
}
