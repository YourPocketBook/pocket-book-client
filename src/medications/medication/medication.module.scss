@import "../../styles/colors.scss";

.background {
  background-color: $color-sja-dark-green;
  min-height: 100vh;
  position: relative;
  padding: 10px;
}

.centreBox {
  background-color: $color-sja-white;
  width: calc(100%-20px);
  border-radius: 5px;
  padding: 10px;
}

.p {
  h1 {
    text-align: center;
    color: $color-sja-dark-green;
  }

  h2 {
    text-align: center;
    color: $color-sja-black;
    font-size: 1.5rem;
  }

  h3 {
    color: $color-sja-black;
    font-size: 1.5rem;
  }

  p,
  li {
    color: $color-sja-black;
    font-size: 1.25rem;
  }
}

footer {
  margin-bottom: 6px;
  color: lighten($color: $color-sja-black, $amount: 50);
}
