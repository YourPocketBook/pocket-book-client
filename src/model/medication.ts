export interface IMedication {
  id: number;
  name: string;
  indications: string;
  inclusionCriteria: string;
  exclusionCriteria: string;
  form: string;
  route: string;
  dose: string;
  sideEffects: string;
  adviceIfTaken: string;
  adviceIfDeclined: string;
  policyDate?: string | null;
}
