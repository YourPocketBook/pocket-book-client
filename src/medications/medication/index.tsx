import format from "date-fns/format";
import parseDate from "date-fns/parseISO";
import React, { useState } from "react";
import Button from "reactstrap/lib/Button";
import ButtonGroup from "reactstrap/lib/ButtonGroup";
import Col from "reactstrap/lib/Col";
import Modal from "reactstrap/lib/Modal";
import ModalBody from "reactstrap/lib/ModalBody";
import ModalFooter from "reactstrap/lib/ModalFooter";
import ModalHeader from "reactstrap/lib/ModalHeader";
import Row from "reactstrap/lib/Row";
import { fetcher } from "../../fetcher";
import { history } from "../../history";
import { useMedication } from "../../hooks/useMedications";
import { useOnlineStatus } from "../../hooks/useOnlineStatus";
import { Link } from "../../link";
import { PrivatePage } from "../../private-page";
import { MarkdownSkeleton } from "./markdown-skeleton";
import styles from "./medication.module.scss";

const deleteEndPoint = "/api/delete-medication";

/**
 * Component that displays the medication to the user and allows administrators
 * to edit or delete medications.
 *
 * Invalid or expired token is handled by [[PrivatePage]].
 *
 * If the user is offline, it displays the offline warning and disables
 * the edit and delete buttons.
 *
 * If load into state returns nothing or fails, it displays an error message.
 *
 * While the delete request is running, it disables the submit button
 * and displays a "busy" indicator.
 * When the delete request completes, it redirects to the medication list.
 */
export function Medication({ id, reload }: { id: string; reload: boolean }) {
  const { medication, isLoading, loadFailed } = useMedication(id, reload);
  const [isBusy, setIsBusy] = useState(false);

  let content;

  const isAdmin = fetcher.isAdmin();
  const isOnline = useOnlineStatus();
  const [showDeleteConfirm, setShowDeleteConfirm] = useState(false);

  function deleteMedication() {
    setIsBusy(true);

    fetcher
      .authFetch(`${deleteEndPoint}/${id}`, {
        method: "DELETE"
      })
      .then(() => {
        setIsBusy(false);
        history.push("/medications");
      });
  }

  if (isLoading && !medication) {
    content = <div>Loading...</div>;
  } else if (!isLoading && (loadFailed || !medication)) {
    content = (
      <React.Fragment>
        <h1>Error Loading Medication</h1>
        <p>We weren't able to load that medication. Try refreshing the page.</p>
      </React.Fragment>
    );
  } else if (medication) {
    content = (
      <React.Fragment>
        <Row>
          <Col>
            <Link
              className="btn btn-link text-decorate-underline"
              href="/medications"
            >
              Go back...
            </Link>
          </Col>
          <Col>
            <h1 className="text-capitalize">
              {(medication.name || "").toLowerCase()}
            </h1>
          </Col>
          <Col />
        </Row>
        <h2>Clinical Condition</h2>
        <h3>Indication</h3>
        <MarkdownSkeleton source={medication.indications} />
        <h3>Inclusion Criteria</h3>
        <MarkdownSkeleton source={medication.inclusionCriteria} />
        <h3>Exclusion Criteria</h3>
        <MarkdownSkeleton source={medication.exclusionCriteria} />
        <h2>Medicine Details</h2>
        <h3>Name, Form and Strength of Medicine</h3>
        <MarkdownSkeleton source={medication.form} />
        <h3>Route or Method</h3>
        <MarkdownSkeleton source={medication.route} />
        <h3>Dosage</h3>
        <MarkdownSkeleton source={medication.dose} />
        <h3>Side Effects</h3>
        <MarkdownSkeleton source={medication.sideEffects} />
        <h3>Advice to Patient and/or Carer</h3>
        <MarkdownSkeleton source={medication.adviceIfTaken} />
        <h3>Advice if Patient Declines or is Excluded</h3>
        <MarkdownSkeleton source={medication.adviceIfDeclined} />
        <h2>Audit Trail</h2>
        <h3>Records</h3>
        <p>
          All details should be recorded on the Patient Report Form and any
          adverse reactions reported via the Incident Management Framework.
        </p>
        <footer>
          {medication.policyDate
            ? `Taken from Medications Administration Directives dated ${format(
                parseDate(medication.policyDate),
                "d MMM yyyy"
              )}.`
            : null}
        </footer>
        {isAdmin ? (
          <ButtonGroup className="text-center">
            <Link
              className="btn btn-dark-green btn-lg"
              href={`/admin/edit-medication/${id}`}
              disabled={!isOnline}
            >
              Edit Medication
            </Link>
            <Button
              className="btn btn-danger btn-lg"
              onClick={() => setShowDeleteConfirm(true)}
              disabled={!isOnline}
            >
              Delete Medication
            </Button>
          </ButtonGroup>
        ) : null}
        <Modal
          isOpen={showDeleteConfirm}
          toggle={() => setShowDeleteConfirm(!showDeleteConfirm)}
        >
          <ModalHeader>Are you sure?</ModalHeader>
          <ModalBody>
            Are you sure you want to delete '{medication.name}'? This cannot be
            un-done.
          </ModalBody>
          <ModalFooter>
            <Button color="danger" onClick={deleteMedication} disabled={isBusy}>
              Yes
            </Button>
            <Button
              color="secondary"
              onClick={() => setShowDeleteConfirm(false)}
            >
              No
            </Button>
          </ModalFooter>
        </Modal>
      </React.Fragment>
    );
  }

  return (
    <PrivatePage>
      <div className={styles.background}>
        <div className={styles.centreBox}>
          <div className={styles.p}>{content}</div>
        </div>
      </div>
    </PrivatePage>
  );
}
