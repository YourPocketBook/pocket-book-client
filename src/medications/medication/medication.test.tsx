import "@testing-library/jest-dom/extend-expect";
import {
  fireEvent,
  render,
  wait,
  waitForDomChange,
  waitForElement
} from "@testing-library/react";
import fetchMock from "fetch-mock";
import React from "react";
import ReactDOM from "react-dom";
import { Medication } from ".";
import {
  authenticatedDelete,
  authenticatedGet
} from "../../_testSupport/fetch-helpers";
import {
  inDateAdminToken,
  inDateNonAdminToken
} from "../../_testSupport/tokens";
import { fetcher } from "../../fetcher";
import { history } from "../../history";
import { useOnlineStatus } from "../../hooks/useOnlineStatus";
import { IMedication } from "../../model/medication";

jest.mock("../../hooks/useOnlineStatus");

const getEndpoint = "/api/get-medication";
const deleteEndPoint = "/api/delete-medication";

const testDefaultState = {
  adviceIfDeclined: "Test Advice 1",
  adviceIfTaken: "Test Advice 2",
  dose: "Test Dose",
  exclusionCriteria: "Test Exclusion",
  form: "Test Form",
  id: 42,
  inclusionCriteria: "Test Inclusion",
  indications: "Test Indications",
  name: "Test Name",
  policyDate: "2018-09-10",
  route: "Test Route",
  sideEffects: "Test Side Effects"
};

describe("Medication Detail Page", () => {
  beforeEach(() => {
    fetcher.saveToken(inDateNonAdminToken);
    history.push = jest.fn();
    (useOnlineStatus as jest.Mock).mockReturnValue(true);
  });

  afterEach(() => {
    localStorage.clear();
    fetcher.clearToken();
  });

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Medication id="42" reload={false} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders cached results correctly", async () => {
    authenticatedGet(
      `${getEndpoint}/42`,
      inDateNonAdminToken,
      // tslint:disable-next-line: no-empty
      new Promise(() => {})
    );

    localStorage.setItem("medications", JSON.stringify([testDefaultState]));

    const { container } = render(<Medication id="42" reload={false} />);

    expect(container).toMatchSnapshot();
  });

  it("renders partial cached results correctly", async () => {
    authenticatedGet(
      `${getEndpoint}/42`,
      inDateNonAdminToken,
      // tslint:disable-next-line: no-empty
      new Promise(() => {})
    );

    const testMedication: Array<Partial<IMedication>> = [
      {
        id: 42,
        name: "Test Name"
      }
    ];

    localStorage.setItem("medications", JSON.stringify(testMedication));

    const { container } = render(<Medication id="42" reload={false} />);

    expect(container).toMatchSnapshot();
  });

  it("updates cached results correctly", async () => {
    authenticatedGet(
      `${getEndpoint}/42`,
      inDateNonAdminToken,
      testDefaultState
    );

    const testMedication: Array<Partial<IMedication>> = [
      {
        id: 42,
        name: "Test Name"
      },
      { id: 43, name: "Test Name 2" }
    ];

    localStorage.setItem("medications", JSON.stringify(testMedication));

    render(<Medication id="42" reload={true} />);

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(JSON.parse(localStorage.getItem("medications") || "")).toEqual([
      testDefaultState,
      { id: 43, name: "Test Name 2" }
    ]);
  });

  it("renders results correctly", async () => {
    authenticatedGet(
      `${getEndpoint}/42`,
      inDateNonAdminToken,
      testDefaultState
    );

    const { container } = render(<Medication id="42" reload={false} />);

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(container).toMatchSnapshot();
  });

  it("renders no policy date result correctly", async () => {
    const noDateResult = { ...testDefaultState, policyDate: null };

    authenticatedGet(`${getEndpoint}/42`, inDateNonAdminToken, noDateResult);

    const { container } = render(<Medication id="42" reload={false} />);

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(container).toMatchSnapshot();
  });

  it("renders results correctly for administrator", async () => {
    fetcher.saveToken(inDateAdminToken);
    authenticatedGet(`${getEndpoint}/42`, inDateAdminToken, testDefaultState);

    const { getByText } = render(<Medication id="42" reload={false} />);

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(getByText("Edit Medication")).toBeInTheDocument();
    expect(getByText("Delete Medication")).toBeInTheDocument();
  });

  it("renders results correctly for administrator when offline", async () => {
    fetcher.saveToken(inDateAdminToken);
    (useOnlineStatus as jest.Mock).mockReturnValue(false);

    localStorage.setItem("medications", JSON.stringify([testDefaultState]));

    const { getByText } = render(<Medication id="42" reload={false} />);

    expect(getByText("Edit Medication")).toHaveClass("disabled");
    expect(getByText("Delete Medication")).toHaveClass("disabled");
  });

  it("makes request to delete medication correctly", async () => {
    fetcher.saveToken(inDateAdminToken);
    authenticatedGet(`${getEndpoint}/42`, inDateAdminToken, testDefaultState);
    authenticatedDelete(`${deleteEndPoint}/42`, inDateAdminToken, {
      status: 204
    });

    const { getByText, container } = render(
      <Medication id="42" reload={false} />
    );

    await waitForDomChange({ container });

    fireEvent.click(getByText("Delete Medication"));

    await waitForElement(() => getByText("Are you sure?"));

    fireEvent.click(getByText("Yes"));

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(history.push).toBeCalledWith("/medications");
  });

  it("renders loading correctly", async () => {
    authenticatedGet(
      `${getEndpoint}/42`,
      inDateNonAdminToken,
      // tslint:disable-next-line: no-empty
      new Promise(() => {})
    );

    const { container } = render(<Medication id="42" reload={false} />);

    expect(container).toMatchSnapshot();
  });

  it("renders failed to load correctly", async () => {
    authenticatedGet(`${getEndpoint}/42`, inDateNonAdminToken, { status: 500 });

    const { container } = render(<Medication id="42" reload={false} />);

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(container).toMatchSnapshot();
  });

  it("renders failed to load correctly when fetch throws", async () => {
    authenticatedGet(`${getEndpoint}/42`, inDateNonAdminToken, {
      throws: new Error()
    });

    const { container } = render(<Medication id="42" reload={false} />);

    await wait(() => expect(fetchMock.done()).toBeTruthy());

    expect(container).toMatchSnapshot();
  });
});
