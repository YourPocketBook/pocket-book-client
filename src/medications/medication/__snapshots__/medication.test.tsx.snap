// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Medication Detail Page renders cached results correctly 1`] = `
<div>
  <div
    class="background"
  >
    <nav
      class="navbar navbar"
    >
      <a
        class="navbar-brand"
      >
        PocketBook
      </a>
      <span
        class="navbar-text"
      >
        Hi, 
        Test User
      </span>
    </nav>
    <div
      class="centreContainer"
    >
      <div
        class="background"
      >
        <div
          class="centreBox"
        >
          <div
            class="p"
          >
            <div
              class="row"
            >
              <div
                class="col"
              >
                <a
                  class="btn btn-link text-decorate-underline"
                  href="/medications"
                >
                  Go back...
                </a>
              </div>
              <div
                class="col"
              >
                <h1
                  class="text-capitalize"
                >
                  test name
                </h1>
              </div>
              <div
                class="col"
              />
            </div>
            <h2>
              Clinical Condition
            </h2>
            <h3>
              Indication
            </h3>
            <p>
              Test Indications
            </p>
            <h3>
              Inclusion Criteria
            </h3>
            <p>
              Test Inclusion
            </p>
            <h3>
              Exclusion Criteria
            </h3>
            <p>
              Test Exclusion
            </p>
            <h2>
              Medicine Details
            </h2>
            <h3>
              Name, Form and Strength of Medicine
            </h3>
            <p>
              Test Form
            </p>
            <h3>
              Route or Method
            </h3>
            <p>
              Test Route
            </p>
            <h3>
              Dosage
            </h3>
            <p>
              Test Dose
            </p>
            <h3>
              Side Effects
            </h3>
            <p>
              Test Side Effects
            </p>
            <h3>
              Advice to Patient and/or Carer
            </h3>
            <p>
              Test Advice 2
            </p>
            <h3>
              Advice if Patient Declines or is Excluded
            </h3>
            <p>
              Test Advice 1
            </p>
            <h2>
              Audit Trail
            </h2>
            <h3>
              Records
            </h3>
            <p>
              All details should be recorded on the Patient Report Form and any adverse reactions reported via the Incident Management Framework.
            </p>
            <footer>
              Taken from Medications Administration Directives dated 10 Sep 2018.
            </footer>
          </div>
        </div>
      </div>
    </div>
    <nav
      class="fixed-bottom footer navbar"
    >
      <span
        class="navbar-text"
      >
        Design Copyright © 
        2020
         Tony Richards.
      </span>
      <ul
        class="nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/privacy"
          >
            Privacy Policy
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/tsandcs"
          >
            Terms of Use
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`Medication Detail Page renders failed to load correctly 1`] = `
<div>
  <div
    class="background"
  >
    <nav
      class="navbar navbar"
    >
      <a
        class="navbar-brand"
      >
        PocketBook
      </a>
      <span
        class="navbar-text"
      >
        Hi, 
        Test User
      </span>
    </nav>
    <div
      class="centreContainer"
    >
      <div
        class="background"
      >
        <div
          class="centreBox"
        >
          <div
            class="p"
          >
            <h1>
              Error Loading Medication
            </h1>
            <p>
              We weren't able to load that medication. Try refreshing the page.
            </p>
          </div>
        </div>
      </div>
    </div>
    <nav
      class="fixed-bottom footer navbar"
    >
      <span
        class="navbar-text"
      >
        Design Copyright © 
        2020
         Tony Richards.
      </span>
      <ul
        class="nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/privacy"
          >
            Privacy Policy
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/tsandcs"
          >
            Terms of Use
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`Medication Detail Page renders failed to load correctly when fetch throws 1`] = `
<div>
  <div
    class="background"
  >
    <nav
      class="navbar navbar"
    >
      <a
        class="navbar-brand"
      >
        PocketBook
      </a>
      <span
        class="navbar-text"
      >
        Hi, 
        Test User
      </span>
    </nav>
    <div
      class="centreContainer"
    >
      <div
        class="background"
      >
        <div
          class="centreBox"
        >
          <div
            class="p"
          >
            <h1>
              Error Loading Medication
            </h1>
            <p>
              We weren't able to load that medication. Try refreshing the page.
            </p>
          </div>
        </div>
      </div>
    </div>
    <nav
      class="fixed-bottom footer navbar"
    >
      <span
        class="navbar-text"
      >
        Design Copyright © 
        2020
         Tony Richards.
      </span>
      <ul
        class="nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/privacy"
          >
            Privacy Policy
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/tsandcs"
          >
            Terms of Use
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`Medication Detail Page renders loading correctly 1`] = `
<div>
  <div
    class="background"
  >
    <nav
      class="navbar navbar"
    >
      <a
        class="navbar-brand"
      >
        PocketBook
      </a>
      <span
        class="navbar-text"
      >
        Hi, 
        Test User
      </span>
    </nav>
    <div
      class="centreContainer"
    >
      <div
        class="background"
      >
        <div
          class="centreBox"
        >
          <div
            class="p"
          >
            <div>
              Loading...
            </div>
          </div>
        </div>
      </div>
    </div>
    <nav
      class="fixed-bottom footer navbar"
    >
      <span
        class="navbar-text"
      >
        Design Copyright © 
        2020
         Tony Richards.
      </span>
      <ul
        class="nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/privacy"
          >
            Privacy Policy
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/tsandcs"
          >
            Terms of Use
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`Medication Detail Page renders no policy date result correctly 1`] = `
<div>
  <div
    class="background"
  >
    <nav
      class="navbar navbar"
    >
      <a
        class="navbar-brand"
      >
        PocketBook
      </a>
      <span
        class="navbar-text"
      >
        Hi, 
        Test User
      </span>
    </nav>
    <div
      class="centreContainer"
    >
      <div
        class="background"
      >
        <div
          class="centreBox"
        >
          <div
            class="p"
          >
            <div
              class="row"
            >
              <div
                class="col"
              >
                <a
                  class="btn btn-link text-decorate-underline"
                  href="/medications"
                >
                  Go back...
                </a>
              </div>
              <div
                class="col"
              >
                <h1
                  class="text-capitalize"
                >
                  test name
                </h1>
              </div>
              <div
                class="col"
              />
            </div>
            <h2>
              Clinical Condition
            </h2>
            <h3>
              Indication
            </h3>
            <p>
              Test Indications
            </p>
            <h3>
              Inclusion Criteria
            </h3>
            <p>
              Test Inclusion
            </p>
            <h3>
              Exclusion Criteria
            </h3>
            <p>
              Test Exclusion
            </p>
            <h2>
              Medicine Details
            </h2>
            <h3>
              Name, Form and Strength of Medicine
            </h3>
            <p>
              Test Form
            </p>
            <h3>
              Route or Method
            </h3>
            <p>
              Test Route
            </p>
            <h3>
              Dosage
            </h3>
            <p>
              Test Dose
            </p>
            <h3>
              Side Effects
            </h3>
            <p>
              Test Side Effects
            </p>
            <h3>
              Advice to Patient and/or Carer
            </h3>
            <p>
              Test Advice 2
            </p>
            <h3>
              Advice if Patient Declines or is Excluded
            </h3>
            <p>
              Test Advice 1
            </p>
            <h2>
              Audit Trail
            </h2>
            <h3>
              Records
            </h3>
            <p>
              All details should be recorded on the Patient Report Form and any adverse reactions reported via the Incident Management Framework.
            </p>
            <footer />
          </div>
        </div>
      </div>
    </div>
    <nav
      class="fixed-bottom footer navbar"
    >
      <span
        class="navbar-text"
      >
        Design Copyright © 
        2020
         Tony Richards.
      </span>
      <ul
        class="nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/privacy"
          >
            Privacy Policy
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/tsandcs"
          >
            Terms of Use
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`Medication Detail Page renders partial cached results correctly 1`] = `
<div>
  <div
    class="background"
  >
    <nav
      class="navbar navbar"
    >
      <a
        class="navbar-brand"
      >
        PocketBook
      </a>
      <span
        class="navbar-text"
      >
        Hi, 
        Test User
      </span>
    </nav>
    <div
      class="centreContainer"
    >
      <div
        class="background"
      >
        <div
          class="centreBox"
        >
          <div
            class="p"
          >
            <div
              class="row"
            >
              <div
                class="col"
              >
                <a
                  class="btn btn-link text-decorate-underline"
                  href="/medications"
                >
                  Go back...
                </a>
              </div>
              <div
                class="col"
              >
                <h1
                  class="text-capitalize"
                >
                  test name
                </h1>
              </div>
              <div
                class="col"
              />
            </div>
            <h2>
              Clinical Condition
            </h2>
            <h3>
              Indication
            </h3>
            <span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
            </span>
            <h3>
              Inclusion Criteria
            </h3>
            <span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
            </span>
            <h3>
              Exclusion Criteria
            </h3>
            <span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
            </span>
            <h2>
              Medicine Details
            </h2>
            <h3>
              Name, Form and Strength of Medicine
            </h3>
            <span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
            </span>
            <h3>
              Route or Method
            </h3>
            <span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
            </span>
            <h3>
              Dosage
            </h3>
            <span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
            </span>
            <h3>
              Side Effects
            </h3>
            <span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
            </span>
            <h3>
              Advice to Patient and/or Carer
            </h3>
            <span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
            </span>
            <h3>
              Advice if Patient Declines or is Excluded
            </h3>
            <span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
              <span
                class="css-19bon8n react-loading-skeleton"
                style="animation: animation-16jpnkj 1.2s ease-in-out infinite;"
              >
                ‌
              </span>
            </span>
            <h2>
              Audit Trail
            </h2>
            <h3>
              Records
            </h3>
            <p>
              All details should be recorded on the Patient Report Form and any adverse reactions reported via the Incident Management Framework.
            </p>
            <footer />
          </div>
        </div>
      </div>
    </div>
    <nav
      class="fixed-bottom footer navbar"
    >
      <span
        class="navbar-text"
      >
        Design Copyright © 
        2020
         Tony Richards.
      </span>
      <ul
        class="nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/privacy"
          >
            Privacy Policy
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/tsandcs"
          >
            Terms of Use
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`Medication Detail Page renders results correctly 1`] = `
<div>
  <div
    class="background"
  >
    <nav
      class="navbar navbar"
    >
      <a
        class="navbar-brand"
      >
        PocketBook
      </a>
      <span
        class="navbar-text"
      >
        Hi, 
        Test User
      </span>
    </nav>
    <div
      class="centreContainer"
    >
      <div
        class="background"
      >
        <div
          class="centreBox"
        >
          <div
            class="p"
          >
            <div
              class="row"
            >
              <div
                class="col"
              >
                <a
                  class="btn btn-link text-decorate-underline"
                  href="/medications"
                >
                  Go back...
                </a>
              </div>
              <div
                class="col"
              >
                <h1
                  class="text-capitalize"
                >
                  test name
                </h1>
              </div>
              <div
                class="col"
              />
            </div>
            <h2>
              Clinical Condition
            </h2>
            <h3>
              Indication
            </h3>
            <p>
              Test Indications
            </p>
            <h3>
              Inclusion Criteria
            </h3>
            <p>
              Test Inclusion
            </p>
            <h3>
              Exclusion Criteria
            </h3>
            <p>
              Test Exclusion
            </p>
            <h2>
              Medicine Details
            </h2>
            <h3>
              Name, Form and Strength of Medicine
            </h3>
            <p>
              Test Form
            </p>
            <h3>
              Route or Method
            </h3>
            <p>
              Test Route
            </p>
            <h3>
              Dosage
            </h3>
            <p>
              Test Dose
            </p>
            <h3>
              Side Effects
            </h3>
            <p>
              Test Side Effects
            </p>
            <h3>
              Advice to Patient and/or Carer
            </h3>
            <p>
              Test Advice 2
            </p>
            <h3>
              Advice if Patient Declines or is Excluded
            </h3>
            <p>
              Test Advice 1
            </p>
            <h2>
              Audit Trail
            </h2>
            <h3>
              Records
            </h3>
            <p>
              All details should be recorded on the Patient Report Form and any adverse reactions reported via the Incident Management Framework.
            </p>
            <footer>
              Taken from Medications Administration Directives dated 10 Sep 2018.
            </footer>
          </div>
        </div>
      </div>
    </div>
    <nav
      class="fixed-bottom footer navbar"
    >
      <span
        class="navbar-text"
      >
        Design Copyright © 
        2020
         Tony Richards.
      </span>
      <ul
        class="nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/privacy"
          >
            Privacy Policy
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-dark"
            href="/tsandcs"
          >
            Terms of Use
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
